docs/operations.rst

```diff
@@ -14,42 +14,169 @@ Provisioning
 
-See `PostgreSQL documentation`__ on setting up `.pgpass` file.
+	When upgrading the monitor, it may happen that this procedure has a new
+	API contract. In that case queries from the running ``pg_autoctl`` are
+	going to fail. We need to upgrade the client program so that we have the
+	code that matches the monitor extension's version.
 
-__ https://www.postgresql.org/docs/current/libpq-pgpass.html
+As a result, here is the standard upgrade plan for pg_auto_failover:
 
+  1. Put all the current secondary nodes in maintenance to prevent from any
+     failover to happen.
 
-Operations
-----------
+	 Run the follow SQL query on the monitor database to make sure your
+	 setup current state allows for an upgrade operation to happen::
+
+		select nodeid, nodename, nodeport, goalstate, reportedstate
+		  from pgautofailover.node
+		 where reportedstate = 'secondary'
+		   and not exists
+			  (
+			   select nodeid
+				 from pgautofailover.node
+				where goalstate <> reportedstate
+				   or reportedstate not in ('primary', 'secondary')
+			  )
+		order by nodeid;
+
+	 The previous SQL query is expected to list all your Postgres nodes that
+	 are currently in the `secondary` state. If the cluster is not in a
+	 globally stable state, then this query returns no rows, and you should
+	 stop the upgrade now.
+
+	 A globally stable state is achieved when all the registered nodes are
+	 either in the state `primary` or in the state `secondary`, and their
+	 currently reported and goal state are the same.
+
+	 Run the follow SQL query on the monitor database to enable the
+	 maintenance operation::
+
+		select nodeid, nodename, nodeport,
+		       pgautofailover.start_maintenance(nodename, nodeport)
+		  from pgautofailover.node
+		 where reportedstate = 'secondary'
+		   and not exists
+			  (
+			   select nodeid
+				 from pgautofailover.node
+				where goalstate <> reportedstate
+				   or reportedstate not in ('primary', 'secondary')
+			  )
+		order by nodeid;
+
+	 Check that the selected nodes have reached `maintenance` by watching
+	 the output of the ``pg_autoctl show state`` command. When all selected
+	 nodes are in maintenance, we can move forward with the upgrade.
+
+	 .. note::
+
+		Alternatively you may also run the command ``pg_autoctl enable
+		maintenance`` on every secondary node.
+
+  2. Stop the ``pg_autoctl`` service on the monitor.
+
+	 The command to stop the systemd pgautofailover service is::
+
+	   sudo systemctl stop pgautofailover
+
+	 To verify that the service has been stopped as expected, we run the
+	 following command::
+
+	   sudo systemctl status pgautofailover
+
+	 At this point all the Postgres nodes in the system are going to error
+	 out every second when connecting to the monitor. In case of error to
+	 connect to the monitor, the node-active process retries every second,
+	 and will be able to continue normally as soon as the monitor is back
+	 online.
+
+  3. Upgrade the pg_auto_failover package on the monitor.
```

Comment:
I thought this step should be after upgrading (and restarting) pg_auto_failover on the postgres nodes. Otherwise the old pg_autoctl keeper processes will communicate with the new monitor extension. Sounds like that could mess stuff up.

Reply:
It's problematic in both cases in different ways. We don't guarantee API compatibility in either direction. 

What I now think we should do, given your feedback here, is to check for the monitor extension version to be as expected by the client each time we open a connection to the monitor, as the first query we do. When the version is not as expected, close the connection and return false to indicate we failed to open it, and let the main loop take over from there.

I believe that if we implement that, then you can upgrade the client pg_autoctl nodes at anytime without consequences: we would only call the API when we're on the expected version anyway, and keep retrying otherwise.

We might even be able to run `pg_autoctl version --json` and parse the result when we see that the server version is different from the current in-memory expected version, and when we see that the on-disk version is the same as the monitor version then just restart automatically?

For that we might want to add the supported monitor version in the JSON output, it's not there yet.

Reply:
Fixed in #318 then #316.